### PR TITLE
Add connection timeout and respawn options

### PR DESCRIPTION
**Pull request by QuentinTorg**
The `camera_timeout_s` will cause the multisense node to shut down if it cannot communicate with the camera for `camera_timeout_s` seconds. Setting this value <=0 will disable the timeout check 

The `respawn` being true means the multisense launch file will attempt to re-launch the node if it is shut down

Used together these two options can be used to automatically re-connect to a camera after it is disconnected/reconnected or power cycled

### CODE REVIEW

File: multisense_ros/src/ros_driver.cpp [hunk 1]
```diff
@@ -121,7 +123,37 @@ int main(int argc, char** argvPP)
...
+            ros::Duration warn_delay{0.5};
+            while (ros::ok()) {
+                ros::spinOnce();
+
+                system::StatusMessage statusMessage;
+                const auto status_result = d->getDeviceStatus(statusMessage);
+                if (Status_Ok != status_result) {
+
+                    if (ros::Time::now() - last_warning > warn_delay) {
+                        ROS_WARN("multisense_ros: lost connection with camera");
```
Comment (mattalvarado) [inline multisense_ros/src/ros_driver.cpp]: This warning is fairly aggressive. Consider softening the language.

File: multisense_ros/src/ros_driver.cpp [hunk 2]
```diff
@@ -121,7 +123,37 @@ int main(int argc, char** argvPP)
...
+                        last_warning = ros::Time::now();
+                    }
+
+                    if (timeout > ros::Duration{0} && ros::Time::now() - last_status > timeout) {
+                        ROS_ERROR("multisense_ros: shutting down due to connection timeout");
```
Comment (mattalvarado) [inline multisense_ros/src/ros_driver.cpp:146]: Consider changing this to something like "reinitializing connection with the MultiSense"

File: multisense_ros/src/ros_driver.cpp [hunk 3]
```diff
@@ -121,7 +123,37 @@ int main(int argc, char** argvPP)
...
+            ros::Time last_status{ros::Time::now()};
+            const ros::Duration timeout{timeout_s};
+
+            ros::Time last_warning{};
+            ros::Duration warn_delay{0.5};
+            while (ros::ok()) {
+                ros::spinOnce();
+
+                system::StatusMessage statusMessage;
+                const auto status_result = d->getDeviceStatus(statusMessage);
```
Comment (mattalvarado) [inline multisense_ros/src/ros_driver.cpp:137]: This will only get updated at 1Hz under the hood.

File: multisense_ros/src/ros_driver.cpp [hunk 4]
```diff
@@ -121,7 +123,37 @@ int main(int argc, char** argvPP)
...
+                if (Status_Ok != status_result) {
+
+                    if (ros::Time::now() - last_warning > warn_delay) {
```
Comment (mattalvarado) [inline multisense_ros/src/ros_driver.cpp:140]: Seems like this might be too chatty when a single status message is dropped

**Comment by QuentinTorg** [comment]
@mattalvarado I have updated so it does not print a warning until the status message has been missing for a full second, and only print at 1Hz to match the status message publish rate. The comment in the launch file for the camera_timeout setting also indicates times of 3s or more should be preferred

File: multisense_ros/src/ros_driver.cpp (hunk 2 above)
Comment (QuentinTorg) [reply multisense_ros/src/ros_driver.cpp:146]: I'm not sure what less harsh comment would be applicable. During shutdown, we only respawn if the respawn argument was also passed, so I don't quite want to say that here. After this comment, we also exit with an error code, so it seems acceptable to be harsh

**Review by mattalvarado — APPROVED** [review]

**Pull request merged**